### ISSUE TITLE
Installation issue: timemory

### ISSUE THREAD

**Issue by mpokorny** [LLNL/spack#27587]
### Steps to reproduce the issue

```console
$ spack install timemory+cuda
```

### Information on your system

* **Spack:** 0.17.0-40-6e6847d9c7
* **Python:** 3.8.12
* **Platform:** linux-rhel7-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/7582343/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/7582344/spack-build-out.txt)

@jrmadsen 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by jrmadsen** [comment]
@mpokorny this is indeed a strange issue. I'm not 100% sure why this is broken. The code causing the errors hasn't been modified whatsoever in well over a year so I'm guessing this issue exists for previous releases too. 

From looking at the errors, my best guess is there might be some macro substitution issue or something. Or maybe the nvcc compiler with the latest GCC is causing an incorrect path through the pre-processor ifdefs. 

Is there any chance you could try recompiling with GCC 10 or less (instead of 11) and/or CUDA 11.4 or less. I built the develop branch recently with CUDA 11.4 and GCC 9 without any issues. And I've built the develop branch with GCC 11 without issue (w/o CUDA though). I don't have a redhat system to test this on but I'll see if I can reproduce on Ubuntu with a docker container

**Comment by mpokorny** [comment]
Build is OK with gcc 9.3.0 on RHEL7 using either CUDA 11.4.0 or CUDA 11.5.0. I can try using gcc 10 if that would help, but I'd need to install it first.

**Comment by jrmadsen** [comment]
@mpokorny Yeah, this is an NVCC bug when parsing the inheritance of a partial template specialization. I was able to reproduce the bug and fix it with https://github.com/NERSC/timemory/pull/237

**Comment by mpokorny** [comment]
Thanks for fixing this so quickly. I'm happy to work around the issue by using gcc 9.3.0 until the `timemory` branch is merged into `develop`.

**Comment by brandongc** [comment]
I'm not sure this issue is still relevant.

The MR with the fix was merged and NERSC/timemory is now a read-only archive.
